Add OptIn metric advisory parameter (#4809)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

Reviewer: @dashpole This section also need a wording about the opt-in advisory?

Author: Updated in https://github.com/open-telemetry/opentelemetry-specification/pull/4809/commits/319c1f725d71d051360715187d6dee44f63a7f8f.  This section was also missing other instrument advisory parameters (e.g. histogram buckets).  Rather than copying all of this, i've re-organized this so that this section is covered above.

  `aggregation_cardinality_limit` value, the `MeterProvider` MUST apply the
  [default aggregation cardinality limit](#metricreader) the `MetricReader` is
  configured with.

> * **Status**: [Development](../document-status.md) - `enabled` (optional): A

Reviewer: this section feels a bit hard to follow. It uses a different style than the other stream configs.

Rewrote to make it easier, by splitting 3 possible values (unset, true, false). Please check if this reads better.

```
**Status**: [Development](../document-status.md) - `enabled`: A boolean
  denoting whether the instrument produces metric data.

Users can provide an `enabled` value, but it is up to their discretion.
Therefore, the stream configuration parameter needs to be structured to
accept an `enabled` value, but MUST NOT obligate a user to provide one.
The parameter MUST distinguish an unset value from `false`.

* If `enabled` is `true`, the `MeterProvider` MUST aggregate the instrument
  as if its [`OptIn`](#instrument-advisory-parameter-optin) advisory
  parameter were `false`.
* If `enabled` is `false`, the `MeterProvider` MUST use the
  [Drop Aggregation](#drop-aggregation), regardless of the `aggregation`
  provided.
* If `enabled` is unset, the
  [`OptIn`](#instrument-advisory-parameter-optin) advisory parameter
  determines whether the instrument is enabled.

```

Author: addressed in b3b8f612

made with an Instrument:

* Determine the `MeterProvider` which "owns" the Instrument.
* If the `MeterProvider` has no `View` registered, or if the Instrument does
  not match any View's instrument selection criteria:
  * **Status**: [Development](../document-status.md) - If the instrument's
    [OptIn advisory parameter](#instrument-advisory-parameter-optin)
    is set to true, use the [Drop Aggregation](#drop-aggregation).

>   * If OptIn is false, take the Instrument and apply the default

Reviewer: we need to distinguish between OptIn is unset, set to true, set to false.

Author: addressed in 9141b79d

made with an Instrument:

* Determine the `MeterProvider` which "owns" the Instrument.

> * If the `MeterProvider` has no `View` registered, or if the Instrument does

Author: Normative tightening inside a Stable section is unflagged. Call it out explicitly in the PR description so approvers evaluate it deliberately, and add a second CHANGELOG line under `### Metrics`, e.g.

    * If both the View and [Instrument advisory
      parameters](#instrument-advisory-parameters) specify the same aspect of
      the [Stream configuration](#stream-configuration), the setting defined by
      the View MUST take precedence over the advisory parameters, except for the
      [OptIn](#instrument-advisory-parameter-optin) advisory parameter which
      requires `enabled` to be set to `true` to be overridden.
  * (**Development**) If `view_matching_mode` is `composable`, and the Instrument could match the instrument selection criteria of one or more Views:
    * Group the matching Views by their configured stream `name`.

>       * A group contains Views that configure the same stream `name` and all

Author: `enabled` is missing from the composable-view merge-aspect list. Add `enabled` to the parenthetical at `sdk.md:528-529` and state the tie-break explicitly.

      * If both the matching Views and [Instrument advisory
        parameters](#instrument-advisory-parameters) specify the same aspect of
        the [Stream configuration](#stream-configuration), the setting defined
        by the Views MUST take precedence over the advisory parameters, except
        for the [OptIn](#instrument-advisory-parameter-optin) advisory parameter
        which requires `enabled` to be set to `true` to be overridden.

Users can configure match-all Views with `enabled=false` or with the
[Drop aggregation](#drop-aggregation), to disable instruments by default.

> #### View examples

Author: No worked example, and no compliance-matrix coverage, for the SDK half of the feature. Add one Python example to `sdk.md:550-617`:.

the advisory parameter should be used. If neither is provided, all attributes
must be retained.

#### Instrument advisory parameter: `OptIn`

**Status**: [Development](../document-status.md)

This advisory parameter applies to all aggregations.

> When an instrument is `OptIn`, the SDK MUST use the

Reviewer: ```suggestion
When an instrument has `OptIn=true`, the SDK MUST use the
```

[Drop Aggregation](#drop-aggregation) by default. If the user has provided an
Aggregation via View(s), that aggregation takes precedence. If the user

> provides the [Default Aggregation](#default-aggregation) using a View, this

Reviewer: ```suggestion
provides the [Default Aggregation](#default-aggregation) using a View or does not provide Aggregation (which results in [Default Aggregation](#default-aggregation)), this
```

Author: Thats a good idea. It does make DropAggregation a bit redundant, but maybe that is OK. It seems like a better UX to set `enabled: false` compared to `Aggregation: DropAggregation` anyways.

I'm less convinced of the naming change for the advisory parameter.  I like that OptIn matches our semantic conventions, and `NewFoo("my.metric", Enabled=false)` seems slightly less clear to me than `NewFoo("my.metric", OptIn=true)`.

> "enables" the instrument with the same behavior as-if the instrument was not

Author: Good point.  This is also the case for features like exponential histograms. See my comment here for my rationale: https://github.com/open-telemetry/opentelemetry-specification/pull/4809#discussion_r2673044311.  Do you think it is worth adding a new aggregation: e.g. EnabledAggregation, which is the same as DefaultAggregation, but never returns DropAggregation?

> "enables" the instrument with the same behavior as-if the instrument was not

Author: The default aggregation section still has no dependency on OptIn.

the advisory parameter should be used. If neither is provided, all attributes
must be retained.

#### Instrument advisory parameter: `OptIn`

**Status**: [Development](../document-status.md)

This advisory parameter applies to all aggregations.

When an instrument has `OptIn=true`, the SDK MUST use the
[Drop Aggregation](#drop-aggregation). If the user sets `enabled=true` on a
View's [Stream configuration](#stream-configuration), this "enables" the

> instrument with the same behavior as-if the instrument was `OptIn=false` --

Reviewer 2: ```suggestion
instrument with the same behavior as-if the instrument was `OptIn=false`,
```

Author: addressed in d90547fe

> the advisory parameter should be used. If neither is provided, all attributes

Author: The general advisory-precedence statement was not updated, leaving three places that say different things. If F2 is fixed as suggested, `sdk.md:1092-1094` needs no carve-out and the whole thing becomes consistent — the `OptIn` subsection at 1125-1137 carries the mapping.

must be retained.

#### Instrument advisory parameter: `OptIn`

**Status**: [Development](../document-status.md)

This advisory parameter applies to all aggregations.

When an instrument has `OptIn=true`, the SDK MUST use the
[Drop Aggregation](#drop-aggregation). If the user sets `enabled=true` on a
View's [Stream configuration](#stream-configuration), this "enables" the
instrument with the same behavior as-if the instrument was `OptIn=false`,
including respecting other advisory parameters. Setting fields other than
`enabled` on the View, including setting the `aggregation`, does not enable
the instrument.

### Instrument enabled

The synchronous instrument [`Enabled`](./api.md#enabled) MUST return `false`
when either:

- **Status**: [Development](../document-status.md) - The [MeterConfig](#meterconfig)
  of the `Meter` used to create the instrument has parameter `enabled=false`.

> - All [resolved streams](#measurement-processing) for the instrument are

Author: `resolved streams` is an undefined term whose link does not land on a definition. Either use "resulting streams" to match `sdk.md:526`/`535`, or define the term once in Measurement processing.

  configured with the [Drop Aggregation](#drop-aggregation).

> Otherwise, it SHOULD return `true`.

Author: `Instrument enabled` Note is now factually wrong for `OptIn` instruments. Amend in place, e.g.
